Automatic & Annotation Configuration

kdubb opened the pull request:
This is a reimplementation of #40 using the new enricher service. All annotations are now named according to #44.

Some points...
* `WeldJunitEnricher` was modified to pass the JUnit `ExtensionContext` to the `enrich` method. The automatic scanning needed to know about the test configuration (specifically the explicit parameter injection setting) and this requires the context.
* `DefaultWeldJUnitEnricher` was removed. It "defaulted" to adding the entire test package; this is exactly the functionality the automatic scanning was trying to stop. In almost all cases automatic scanning should yield the same results without adding bean classes that are not required. The reason adding the entire test package is not helpful is because some of our tests need mocks, while others need real implementations of the same class; narrowing the included bean classes means each test can easily choose what is included without needing to first exclude certain implementations.
* Fixed a bug in `WeldJUnit5Extension` where it wasn't using the builder provided to the enrichers to create the initiator

manovotn commented:
Thanks for the PR, we'll try to review this throughout the week and add some comments as we glance at it.

> WeldJunitEnricher was modified to pass the JUnit ExtensionContext to the enrich method.

If this is needed, then it can be passes over to `enrich` method instead of `Object testInstance` as that one can be obtained from context anyway.

> DefaultWeldJUnitEnricher was removed. It "defaulted" to adding the entire test package;

Don't like this, we cannot just change default behaviour we had until now. That would be backward incompatible. I was thinking they could both co-exist. I thought that when you want to use automagic, that's basically using special beans you added into same package + some extra from "outside" which the automagic adds. Why else would you add beans to same package and then not use then in tests?

> Fixed a bug in WeldJUnit5Extension...

Nice catch. Though it looks like a design flaw we made, we can either use `Weld` object or builder but never both. I mean user could temper with `Weld` now but unless it is set into the builder object, it won't be used. We should perhaps only pass the builder instance and make `Weld` instance accessible from there? I'll think about it some more...

manovotn commented on junit5/src/main/java/org/jboss/weld/junit5/auto/AddBeanClasses.java:
```diff
@@ -0,0 +1,21 @@
+@Retention(RetentionPolicy.RUNTIME)
+@Target(ElementType.TYPE)
+@Repeatable(AddBeanClasses.All.class)
+public @interface AddBeanClasses {
```
All the annotations need javadoc

manovotn commented on junit5/src/main/java/org/jboss/weld/junit5/auto/TestExtension.java:
```diff
@@ -0,0 +1,65 @@
+
+
+
+public class TestExtension implements Extension {
```
We need to come up with better name for this extension.
The only thing I can think of now is TestInstanceWrapper but that's not good either.

kdubb replied:
I chose `TestInstanceInjectionExtension` as that is exactly what it is for.  I've alos

manovotn commented on junit5/src/test/java/org/jboss/weld/junit5/auto/extension/AddedExtension.java:
```diff
@@ -14,20 +14,28 @@
- * Adds all classes from the test class package.
  *
- * @author Martin Kouba
+ * @author <a href="mailto:manovotn@redhat.com">Matej Novotny</a>
```
How did I become author of this? :-D

kdubb replied:
Copy and paste 😬. Being removed.

mkouba commented on junit5/src/main/java/org/jboss/weld/junit5/WeldJunitEnricher.java:
```diff
@@ -38,10 +41,11 @@
      * @param testInstance
      */
-    void enrich(Weld weld, WeldInitiator.Builder weldInitiatorBuilder, Object testInstance);
+    void enrich(ExtensionContext context, Weld weld, WeldInitiator.Builder weldInitiatorBuilder, Object testInstance);
```
I think we can't, see also `org.junit.jupiter.api.extension.TestInstancePostProcessor.postProcessTestInstance(Object, ExtensionContext)` - .

manovotn replied:
Good point

mkouba replied:
I propose to change the method signature to:
`void enrich(Object testInstance, ExtensionContext context, Weld weld, WeldInitiator.Builder weldInitiatorBuilder)`

mkouba commented on junit5/src/main/java/org/jboss/weld/junit5/auto/AutoWeldJUnitEnricher.java:
```diff
@@ -0,0 +1,30 @@
+
+    weld.addAlternativeStereotype(OverrideBean.class);
+
+    ClassScanning.scanForRequiredBeanClass(testClass, weld, getExplicitInjectionInfoFromStore(context));
```
`ClassScanning` looks like a helper class already...

manovotn replied:
I meant the call to `getExplicitInjectionInfoFromStore(context)` from our junit extension class

kdubb replied:
I will move all this `ExtensionContext` access functionality into a utility class that everybody can use.

mkouba commented on junit5/src/main/java/org/jboss/weld/junit5/auto/AddClasspaths.java:
```diff
@@ -0,0 +1,21 @@
+@Retention(RetentionPolicy.RUNTIME)
+@Target(ElementType.TYPE)
+@Repeatable(AddClasspaths.All.class)
+public @interface AddClasspaths {
```
Pls remove this functionality, we're not going to introduce another classpath magic.

kdubb replied:
This seems to be shortsighted. This functionality is *_very_* useful when adding bean classes from an external library. The only alternative is to use `AddPackages` and interrogate the library to ensure you list all the packages from the library; which may change with each release. For example when using the DeltaSpike library we had to list about 10 different packages (possibly some private) to use the library and this annotation replaces that with a reference to a single public class from DeltaSpike.

I would urge we leave this in. As I have stated this implementation comes from a large library of real world unit tests and it is very useful. We could discuss what other "classpath magic" discovery methods are possible.

mkouba commented:
Few quick notes:
- WeldJunit5Extension bug -good catch and thanks! :+1:
- `DefaultWeldJunitEnricher` must be retained
- `@OverrideBean` - this one really needs a better name, maybe something like  `@SelectedAlternative/@EnabledAlternative/@MockAlternative`
- `TestExtension` - I'm still not convinced we want this workaround to support mock producers taking state from the test instance, but if we decide to keep it a better name must be choosen

Thanks again!

kdubb commented:
@mkouba I've implemented most of that which is requested.  The outstanding issues are...

* `@OverrideBean` was already previously named `@EnabledAlternative`; the suggestion now is to rename it again, possibly back to its old name.  I personally like having the word "override" in the name because it does that and not exactly just enables it as an alternative. Although, either name is fine with me.
* `DefaultWeldJunitEnricher` has not been reinstated. The `automagic` scanning provided by `ClassScanning` _should_ be backwards compatible while selecting a smaller set of classes. Not selecting the test package by default allows us to create test utility that have their own injection and we can control what gets injected via the normal extensions (e.g. `AddBeanClasses`). 

Just a note...
`TestExtension` is not just to support mock producers it handles _all_ the injection for the test instance. As noted above, It has been renamed and can be replaced when the required functionality exists in JUnit 5.

mkouba commented:
@kdubb thanks for updates!

- `@AddClasspaths` - I insist on removing this feature, it's out of the project scope, if a similar functionality is really needed, then use Arquillian Weld Embedded + ShrinkWrap instead (see for example Weld test suite)
- `DefaultWeldJunitEnricher`
  - this change is not entirely backward compatible - injection points analysis does not work if you're programming to the interface (e.g. if there is `Foo` interface and `FooImpl` class, `@Inject Foo` will not add the bean class)
  - also any enricher can be disabled through a system property (see also https://github.com/weld/weld-junit/blob/master/junit5/src/main/java/org/jboss/weld/junit5/WeldJunitEnricher.java#L26)
  - pls bring the class back
- `@OverrideBean`
  - I think the annotation target should be only METHOD and FIELD
  - `@EnabledAlternative` is probably good ...
- _"TestExtension is not just to support mock producers, it handles all the injection for the test instance."_
  - not really, the test instance is first injected here: https://github.com/weld/weld-junit/blob/master/junit5/src/main/java/org/jboss/weld/junit5/WeldInitiator.java#L191
  - in fact, I think there is a problem with the current impl: whenever a producer method is invoked (i.e. the bean instance is needed), the test instance is "reinjected" and this might cause serious problems if a test class injects some `@Dependent` beans
  - still the functionality is probably useful, a simple fix (which should be enough for now) is to add/test `AtomicBoolean isInjected` so that the target instance is only injected once

kdubb commented:
@manovotn @mkouba I'm wondering if this code belongs in this project at this point. My reason for creating this PR was because [CDI-Unit](http://bryncooke.github.io/cdi-unit) has few maintainers, overlaps in functionality and seems to lag behind Weld and JUnit. I thought since the goals were similar enough that we could essentially merge the major functionality into a "Weld" sanctioned project and everybody be better off.

Unfortunately there seems to be a bit of impedance mismatch. You seem to have some objections which I do not understand (e.g. `AddClasspaths`) and for a fledgling project you seem very concerned about backwards compatibility (i.e. `DefaultWeldJUnitEnricher`). 

When it comes to `AddClasspaths` I've provided _*real world*_ examples of why it makes our tests easier to write and easier to maintain; this functionality was modeled after that in CDI-Unit, so other _*real world*_ users find it useful as well. Your answer to this was to tell me to "use something else" because it doesn't belong here. That sounds like dogma not a valid reason.

For all intents and purposes this project is _fairly_ new and the default behavior of including the entire test package is problem in our large test suite; and so I would assume others. It causes us to have to add some mocks and real implementation beans  to _every single test class_ even when a specific class doesn't use the all the objects referenced; simply because by default you include every single bean in the test package in the booted container. The `automagic` scanning is designed it cut the included beans down to only those needed by each test class.  As I stated this _*should*_ be mostly backwards compatible, is it 100% backwards compatible probably not. With a project in a fairly early stage the level of compatibly it provided seems agreeable to me; apparently it is not enough for you.

I tried to use the project "as is" in the last version. In our codebase it meant _*manually*_ adding weld initiator fields to _*hundreds*_ of test classes and then trying to figure out every single bean dependency for each class and adding it to the initiator. Not only is this very "JUnit 4" style, the level of work is basically untenable when compared to other methods. With the changes in this PR we were able to do it by adding a single annotation to each test class (which we already had to do for JUnit 5).  It seems to me one of these methods works really well while the other is very cumbersome.

Before I spend anymore time working on the PR I guess you need to decide the real direction of the project going forward. If it's something like this PR awesome but then you need to open up your scope to include things you may not agree with immediately.  If these things are to different for you, so be it, I'll invest my time in another project that works for me.

To be clear I'm not upset or trying to demean anybody. I'm just trying to make sure my efforts result in a test framework _*we*_ can use.  I don't have time to build a test framework for others and then figure out how to solve our problems separately.

kdubb commented:
@manovotn FYI, Travis runs Maven commands on its own before running the commands specified in `.travis.yml` by auto-activating the profiles based on environment variables it ensures that the selected profiles are activated for those commands as well.

mkouba commented:
Nobody questions the usefulness of most functionalities included in this PR. However, we're not going to merge anything we don't think is entirely correct, nor sacrify backward compatibility ("fledgling" project or not). After all, this is not CDI-Unit and our goal is not to take over its heritage. Nor is our goal to follow the "take all or nothing" PR strategy. In most open source projects I know the maintainers cherry-pick the parts which fit well and represent no stability risks. Anyway, that's the reason why we expose SPI so that anyone can extend the core functionality.

`AddClasspaths` - I can see more risks than benefits here. E.g. if you add a bean archive which itself has beans.xml. This beans.xml would be happily ignored. So it's a no-go.

> Your answer to this was to tell me to "use something else" because it doesn't belong here. That sounds like dogma not a valid reason.

Dogma? Well, I tried to say: use a hammer to drive a nail instead of taking a screwdriver and trying to attach boulder to it to reach similar functionality.

I understand that you feel concerned about the scope of this project, but the reality it that this project is not tailored to a single user needs. It's the community which shapes its future. So it's up to you whether you take the risk. And we hope you will ;-)

kdubb commented:
@mkouba After messing around with the implementation and trying a couple different methods I couldn't really figure a way to make auto configuration and the default behavior play well together.  I now think the best option is to separate them. I implemented that in the latest commit.

I created a new extension `WeldJunit5AutoExtension` and a companion annotation `@EnableAutoWeld`.  This new extension derives from `WeldJunit5Extension` but implements the auto configuration without adding any default packages. 

The nice part about this is people can easily choose between implementations without having to set one or more system properties; essentially they can choose `@EnableWeld` or `@EnableAutoWeld` and it determines the "default" behavior. 

Both implementations kept the new enrichers so they can easily be extended by implementing the `WeldJunit5Enricher` service as before.

Also, I removed the `@AddClasspaths` functionality as requested.

Let me know what everybody thinks of this approach. I think it's a pretty good compromise and makes it very easy for users to choose their preferred default functionality.

manovotn commented:
Hey, good to see you back!
This does sounds like a reasonable compromise to me.
BTW in the current state, the `DefaultWeldJunitEnricher` is completely useless as you never invoke it anyway (and the functionality is now in the extension itself). Plus some other minor changes, but yea, in general, this looks feasible. Let's see what Martin has to say.

kdubb commented:
Yes, there are some cleanup items that need to take place before this would be considered ready. I wanted to get everybody's take on it before spending that time.

mkouba commented:
I like the idea of `@EnableAutoWeld`: +1

kdubb commented on junit5/src/main/java/org/jboss/weld/junit5/ExtensionContextUtils.java:
```diff
@@ -0,0 +1,100 @@
+/**
+ * Utility methods for fetching/storing items stored in the JUnit extension context.
+ */
+public class ExtensionContextUtils {
```
The automatic extension is implemented in the "auto" sub-package; requiring this to be public.

manovotn replied:
Yes, you are right. My previous comment pre-dates the existence of the separate extension hence it's now outdated.

kdubb commented on junit5/src/main/java/org/jboss/weld/junit5/ExtensionContextUtils.java:
```diff
@@ -0,0 +1,100 @@
+    private static final String WELD_ENRICHERS = "weldEnrichers";
+
+    private static final Namespace EXTENSION_NAMESPACE = Namespace.create(WeldJunit5Extension.class);
+    private static Namespace TEST_NAMESPACE;
```
👍

kdubb commented on junit5/src/main/java/org/jboss/weld/junit5/auto/ActivateScopes.java:
```diff
@@ -0,0 +1,31 @@
+ * Activates the listed scopes for the duration of the test.
+ * <p>
+ * This annotation can be placed on a test method, class or suite, although
+ * each scope is activated individually for each test.
```
Description updated.  Hopefully we can implement per-method instantiation in the future and make this true.

kdubb commented on junit5/src/main/java/org/jboss/weld/junit5/auto/TestInstanceInjectionExtension.java:
```diff
@@ -0,0 +1,71 @@
+ * Extension class that ensure proper injection of an externally provided
+ * test instance.
+ *
+ * NOTE: When JUnit 5 provides a test instance creation extension point this
```
👍

kdubb commented on junit5/src/main/java/org/jboss/weld/junit5/auto/TestInstanceInjectionExtension.java:
```diff
@@ -0,0 +1,71 @@
+
+                @Override
+                public void inject(T t, CreationalContext<T> creationalContext) {
+                    wrapped.inject(t, creationalContext);
```
The extension now ensures the test class has a scope of `@Singleton` to ensure this cannot happen.

manovotn replied:
Alright, that should work as well.

kdubb commented on junit5/src/main/java/org/jboss/weld/junit5/auto/OverrideBean.java:
```diff
@@ -0,0 +1,23 @@
+/**
+ * Overrides a bean class that <i>may</i> otherwise be included in the container.
+ *
+ * Using this annotation provided an easy way to replace a bean class during a
```
👍

kdubb commented on junit5/src/main/resources/META-INF/services/org.jboss.weld.junit5.WeldJunitEnricher:
```diff
@@ -1 +0,0 @@
-org.jboss.weld.junit5.DefaultWeldJunitEnricher
```
👍

kdubb commented on junit5/src/main/java/org/jboss/weld/junit5/WeldJunit5Extension.java:
```diff
@@ -159,23 +156,27 @@ public void postProcessTestInstance(Object testInstance, ExtensionContext contex
+        setContainerToStore(context, initiator.initWeld(testInstance));
+    }
+
+    protected void weldInit(Object testInstance, ExtensionContext context, Weld weld, WeldInitiator.Builder weldInitiatorBuilder) {
```
It's already gone.

kdubb commented:
@manovotn I think all your issues have been addressed.

manovotn commented:
@kdubb merged, thanks for contribution!
We'll be doing a release soon to roll this out (within, say, next two weeks).
